airbyte-integrations/connectors/source-mysql/src/test/java/io/airbyte/integrations/source/mysql/CdcMysqlSourceTest.java

```diff
@@ -0,0 +1,446 @@
+  }
+
+  @Override
+  protected void assertCdcMetaData(JsonNode data, boolean deletedAtNull) {
+    assertNotNull(data.get(CDC_LOG_FILE));
+    assertNotNull(data.get(CDC_LOG_POS));
+    assertNotNull(data.get(CDC_UPDATED_AT));
+    if (deletedAtNull) {
+      assertTrue(data.get(CDC_DELETED_AT).isNull());
+    } else {
+      assertFalse(data.get(CDC_DELETED_AT).isNull());
+    }
+  }
+
+  @Override
+  protected void removeCDCColumns(ObjectNode data) {
+    data.remove(CDC_LOG_FILE);
+    data.remove(CDC_LOG_POS);
+    data.remove(CDC_UPDATED_AT);
+    data.remove(CDC_DELETED_AT);
+  }
+
+  @Override
+  protected void addCdcMetadataColumns(AirbyteStream stream) {
+    ObjectNode jsonSchema = (ObjectNode) stream.getJsonSchema();
+    ObjectNode properties = (ObjectNode) jsonSchema.get("properties");
+
+    final JsonNode numberType = Jsons.jsonNode(ImmutableMap.of("type", "number"));
+
+    final JsonNode stringType = Jsons.jsonNode(ImmutableMap.of("type", "string"));
+    properties.set(CDC_LOG_FILE, stringType);
+    properties.set(CDC_LOG_POS, numberType);
+    properties.set(CDC_UPDATED_AT, stringType);
+    properties.set(CDC_DELETED_AT, stringType);
+  }
+
+  @Override
+  protected Source getSource() {
+    return source;
+  }
+
+  @Override
+  protected JsonNode getConfig() {
+    return config;
+  }
+
+  @Override
+  protected Database getDatabase() {
+    return database;
+  }
+
+  @Override
+  public void assertExpectedStateMessages(List<AirbyteStateMessage> stateMessages) {
+    for (AirbyteStateMessage stateMessage : stateMessages) {
+      assertNotNull(stateMessage.getData().get("cdc_state").get("state").get(MYSQL_CDC_OFFSET));
+      assertNotNull(stateMessage.getData().get("cdc_state").get("state").get(MYSQL_DB_HISTORY));
+    }
+  }
+
+  @Override
+  protected AirbyteCatalog expectedCatalogForDiscover() {
+    final AirbyteCatalog expectedCatalog = Jsons.clone(CATALOG);
+
+    createTable(MODELS_SCHEMA, MODELS_STREAM_NAME + "_2",
+        columnClause(ImmutableMap.of(COL_ID, "INTEGER", COL_MAKE_ID, "INTEGER", COL_MODEL, "VARCHAR(200)"), Optional.empty()));
+
+    List<AirbyteStream> streams = expectedCatalog.getStreams();
+    // stream with PK
+    streams.get(0).setSourceDefinedCursor(true);
+    addCdcMetadataColumns(streams.get(0));
+
+    AirbyteStream streamWithoutPK = CatalogHelpers.createAirbyteStream(
+        MODELS_STREAM_NAME + "_2",
+        MODELS_SCHEMA,
+        Field.of(COL_ID, JsonSchemaType.NUMBER),
+        Field.of(COL_MAKE_ID, JsonSchemaType.NUMBER),
+        Field.of(COL_MODEL, JsonSchemaType.STRING));
+    streamWithoutPK.setSourceDefinedPrimaryKey(Collections.emptyList());
+    streamWithoutPK.setSupportedSyncModes(List.of(SyncMode.FULL_REFRESH));
+    addCdcMetadataColumns(streamWithoutPK);
+
+    streams.add(streamWithoutPK);
+    expectedCatalog.withStreams(streams);
+    return expectedCatalog;
+  }
+
+  //TODO : Enable this test once we fix handling of DATETIME values
```

Comment:
which issue is this waiting for? https://github.com/airbytehq/airbyte/issues/14628 ?

Reply:
yup